feat: add & remove to collection

m-m-mic opened the pull request:
# THIS PR IS NOW READY

An updated description can be found [HERE](https://github.com/m-m-mic/Me-My-Shelf-and-I/pull/3#issuecomment-1727684289).

The description below is **outdated**.
___

Hey there,

this is currently a work-in-progress PR in which I'm implementing the ability for users to add and remove media (games, movies & albums) to their personal collections. After they added an item to their collection they can also adjust personal details in regards to said item (e.g. whether they own it physically or digitally, personal notes etc.).

So far I've implemented this for games, but not yet for movies & albums. I think it would be a good idea to go over the existing code now so I can adapt it for movies & albums later. This should also speed up the final review.

**What is currently implemented:**
- Header with links to pages while signed in
- Dashboard now shows games in your collection
- Games, movies and albums pages - games page shows all games in database
- Individual game, movie and album pages - game page shows info and personal details to game based on gameId param in url
- Ability to add and remove game to collection
- Ability to change personal details of a game that's in your collection

![grafik](https://github.com/m-m-mic/Me-My-Shelf-and-I/assets/85286401/7b027fd3-68a4-4e1f-982e-15221db1ed4f)
![grafik](https://github.com/m-m-mic/Me-My-Shelf-and-I/assets/85286401/1485bb5e-beac-47c8-8f4d-eb3d77ff2e86)
![grafik](https://github.com/m-m-mic/Me-My-Shelf-and-I/assets/85286401/09bc9e01-2f2b-4a23-ad5b-d36ec2d9641b)

Most of these changes/additions can be found in the `GamesComponent`, `GameComponent`, `DashboardComponent`, `GamesService` & `UserService`.

I'm still having a hard time writing RxJs. I often end up with several nested subscriptions which become very long and hard to read. If you have ideas how to improve and simplify these RxJs chains I'd be happy to know!

**What is currently missing**:
- Implementation of movies & albums overview pages
- Implementation of individual movie & album pages
- Error handling
- Improved styling

**What will NOT be part of the PR**:
This PR won't include the ability to add games, movies, or albums to the database itself, as I haven't decided yet if and in what capacity regular users should be allowed to add to so.

m-m-mic committed:
feat: add overview page and tab selection to header, add games service and save/remove game to user collection functions

m-m-mic committed:
feat: rename overview to games page, add singular game page, add mmsai-card, add backend getAll and getGame requests to gamesService

m-m-mic committed:
chore: change redirects and text

m-m-mic committed:
chore: add styling, update game types and functions

m-m-mic committed:
feat: add to collection, remove to collection and update data functions to game page

m-m-mic committed:
feat: add update on change to game component, overhaul dashboard page

m-m-mic committed:
chore: update not found page

m-m-mic committed:
feat: add movie and album pages, update interfaces, add saved by users metric

m-m-mic committed:
chore: rename media prop to format, clean up game component

m-m-mic committed:
feat: improve styling for game page

m-m-mic committed:
feat: make header responsive

georgiee commented on src/app/core/components/mmsai-card/mmsai-card.component.ts:
```diff
@@ -0,0 +1,32 @@
+export class MmsaiCardComponent implements OnInit {
+  @Input() data!: GameWithIdType;
+  @Input() cardType!: MediaType;
+  url!: string;
```
maybe better?

```suggestion
  url?: string;
```

georgiee commented on src/app/core/components/mmsai-card/mmsai-card.component.ts:
```diff
@@ -0,0 +1,32 @@
+})
+export class MmsaiCardComponent implements OnInit {
+  @Input() data!: GameWithIdType;
+  @Input() cardType!: MediaType;
```
required is avaialble sicne angualr v16
https://netbasal.com/from-good-to-great-required-inputs-in-angular-47374c8a43cf

the new transform came to my mind, but maybe not?

georgiee commented on src/app/core/layout/header/header.component.ts:
```diff
@@ -8,11 +8,17 @@ import { Store } from '@ngrx/store';
   standalone: true,
-  imports: [CommonModule, RouterLink, MmsaiAccountButtonComponent],
+  imports: [
+    CommonModule,
```
`NgIf` is a subset of common module to import.

georgiee commented on src/app/core/services/games.service.ts:
```diff
@@ -0,0 +1,77 @@
+  getGame(id: string) {
+    return this.gamesRef.doc(id).valueChanges() as Observable<
+      GameType | undefined
+    >;
```
check if there is a generic to do sth like that 
` return this.gamesRef.doc<GameType>(id).valueChanges()`

georgiee commented on src/app/core/services/games.service.ts:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  addUserToGame(gameId: string, userId: string) {
+    this.getGame(gameId)
```
await game = getGame
game...

via 
firstValueFrom
https://rxjs.dev/api/index/function/firstValueFrom

georgiee replied:
maybe wrong direction. if you;'re at a game detail page, why isn't there a "current game" already. Like `currentGame$` or maybe even a resolve one in the store (ngrx via effect)

georgiee commented on src/app/features/dashboard/dashboard.component.ts:
```diff
@@ -1,24 +1,60 @@
+          this.usersService
+            .getUser(user.uid)
+            .pipe(takeUntilDestroyed(destroyRef))
+            .subscribe((value) => {
```
+ two or more subsribes are a red flag

here: switchMap is your friend.


```
    this.authenticationService
      .getUser()
      .pipe(
         take(1),
         switchMap(userAuth => this.usersService.getUser(user.uid)),
        
)), map(user =>
```

georgiee replied:
hints
```
 games$ = this.usersService
            .getUser(user.uid)
            .pipe(
              map(user.games => {
                return value.collection.games.map(game => async {
                    await result = game.ref.get()
                })
              })
            )
```

georgiee commented on src/app/features/dashboard/dashboard.component.ts:
```diff
@@ -1,24 +1,60 @@
+                        user: game,
+                      }),
+                    );
+                  }
```
```
                this.games = value.collection.games.map(...)
```

georgiee commented on src/app/features/game/game.component.html:
```diff
@@ -0,0 +1,54 @@
+    label="Add to your collection"
+    (click)="addToCollection()"></p-button>
+  <p-button
+    *ngIf="inUserCollection"
```
```
<div *ngIf="isLoggedIn; else loggedOut">
  Welcome back, friend.
</div>

<ng-template #loggedOut>
  Please friend, login.
</ng-template>
```


via https://ultimatecourses.com/blog/angular-ngif-else-then

georgiee commented on src/app/features/game/game.component.html:
```diff
@@ -0,0 +1,54 @@
+    <h2>Your Details</h2>
+    <div class="form-selection">
+      <span class="form-container">
+        <h3>Media form:</h3>
```
hierarchy 👍
accessibility (a11y) is very important. wcag 2 is the standard https://www.w3.org/WAI/standards-guidelines/wcag/

there are checklist https://www.a11yproject.com/checklist/

you get the feeling for 80% very easily if you invest some time. take your time 🙌

```
<img src="elephant.jpg" alt="Elephant stands in frotn of four trees ..." />
<img src="http://www.separator.jpg" alt="" />
```

georgiee commented on src/app/features/game/game.component.ts:
```diff
@@ -0,0 +1,121 @@
+  }
+
+  updateData() {
+    if (this.userGameData) {
+      this.usersService.updateGameFromUser(
+        this.uid,
+        createGameObject(this.userGameData.ref, this.gameForm),
+      );
+    }
```
shortcutting is your friend
```suggestion
if (!this.userGameData) {
    return
}

this.usersService.updateGameFromUser(
    this.uid,
    createGameObject(this.userGameData.ref, this.gameForm),
);

```

makes it much more readable.

georgiee commented on src/app/features/game/game.component.ts:
```diff
@@ -0,0 +1,121 @@
+      });
+  }
+
+  getUserGameData() {
+    this.authenticationService
+      .getUser()
+      .pipe(take(1))
+      .subscribe((user) => {
+        if (user) {
+          this.uid = user.uid;
+          this.usersService
+            .getUser(this.uid)
+            .pipe(takeUntilDestroyed(this.destroyRef))
+            .subscribe((value) => {
+              if (value) {
+                this.inUserCollection = false;
+                for (const game of value.collection.games) {
+                  if (game.ref.id === this.id) {
+                    this.inUserCollection = game.in_collection;
+                    this.userGameData = game;
+                    break;
+                  }
+                }
+                this.gameForm = this.formBuilder.group(
+                  createGameForm(this.inUserCollection, this.userGameData),
+                );
+              }
+            });
+        } else {
+          throwError(() => new Error('Could not find user'));
+        }
+      });
+  }
```
that would be better in a service.
"api driven thinking" I would lvoe to have sth like this:

game$ = gameServie.getGame(id)

<game-presentation game="game$ | async"/>

georgiee commented on src/app/features/game/game.component.ts:
```diff
@@ -0,0 +1,121 @@
+                    break;
+                  }
+                }
+                this.gameForm = this.formBuilder.group(
+                  createGameForm(this.inUserCollection, this.userGameData),
+                );
```
this.gameForm = createGameForm()
constructor() {
}


createGameForm = () => {
cosnt formBuodler = inject(FromFBuilder)
return formBuodler.create(...)
}


later:
this.gameForm.setValies({name: ''})

this decouples form build from the hydration (initial) and other updates

you can use the form before the hydration.

this.gameForm.valueChanges.subscribe..

georgiee commented on src/app/features/game/game.items.ts:
```diff
@@ -0,0 +1,23 @@
+export const gameItems = {
```
prime ng requirement. willingly ignoring this

georgiee commented on src/app/features/not-found/not-found.component.html:
```diff
@@ -1 +1,2 @@
-<p>not-found works!</p>
+<h1>Page not found :(</h1>
+<p-button (click)="router.navigate(['/'])" label="Return to Home"></p-button>
```
use routerLink? double check if possible even with ng prime.

georgiee reviewed (commented):
here some comments from our peer review

m-m-mic committed:
feat: make game component responsive

m-m-mic committed:
chore: refactor component names

m-m-mic committed:
chore: refactor and improve game page

m-m-mic committed:
chore: clean up components, rename types

m-m-mic committed:
feat: improve styling of game card, add loading animation and conditions to dashboard

m-m-mic committed:
feat: add moviesService and movie methods to usersService, add logic to movies page

m-m-mic committed:
feat: expand album types

m-m-mic committed:
fix: change import paths in scss files

m-m-mic committed:
feat: add album card, service and user methods

m-m-mic committed:
feat: add movie page, more loading animations

m-m-mic committed:
fix: filter media which isn't in user collection

m-m-mic committed:
feat: add album page

m-m-mic committed:
chore: refactor services and fetch requests to simplify media pages

m-m-mic committed:
chore: separate media overview pages into presentation and container

m-m-mic committed:
chore: delete unused tests, clean up components

m-m-mic commented:
Hey Georgios,

I went through your feedback and implemented the remaining missing pages. The PR is now ready for review.

This PR adds the ability for users to look at media items (games, movies & albums) which are stored in the database and add or remove them to their personal collection. Users can also add personal details to media items in their collection (e.g. whether they own it physically or digitally etc.). 

To accomplish this, I added the following components/services:
- Header with links for page navigation when signed in
- Dashboard page which shows all media items which are currently in the user's collection
- Overview pages for games, movies & albums - Allows user to see what media items are available in the database
- Individual pages for games, movies & albums - Shows more info about the media item in question and allows the user to add or remove the media item to their collection and add personal details
- Game, movie & album cards which function as links to the individual media pages
- Games, movies & albums services which handle database requests from the individual and overview media pages

![20230920_145141](https://github.com/m-m-mic/Me-My-Shelf-and-I/assets/85286401/daa44a9c-a64c-4b57-9993-646f9dccf28c)

The GIF above shows me selecting a game and adding it to my collection, which in turn shows up on my dashboard.

___

This PR contains a lot of redundant code between the three types of media items (Games, movies & albums). Because of this I've left comments below which clarify what components/services you can gloss over while reviewing.

I apologize for the size of this PR. Now that the key features are implemented I'm going to start working on smaller features which should be easier to review. I've also decided to implement proper error handling and improved responsive styling for the added components in a future PR.

Thanks for the review!

m-m-mic committed:
chore: remove unused imports

m-m-mic commented on src/app/core/components/game-card/game-card.component.html:
Same as `album-card.component.html`.

m-m-mic commented on src/app/core/components/game-card/game-card.component.scss:
Same as `album-card.component.scss`

m-m-mic commented on src/app/core/components/game-card/game-card.component.ts:
Same as `album-card.component.ts`

m-m-mic commented on src/app/core/components/movie-card/movie-card.component.html:
Same as `album-card.component.html`

m-m-mic commented on src/app/core/components/movie-card/movie-card.component.scss:
Same as `album-card.component.scss`

m-m-mic commented on src/app/core/components/movie-card/movie-card.component.ts:
Same as `album-card.component.ts`

m-m-mic commented on src/app/core/models/game.interface.ts:
Same as `album.interface.ts`

m-m-mic commented on src/app/core/models/movie.interface.ts:
Same as `album.interface.ts`

m-m-mic commented on src/app/core/services/games.service.ts:
Same as `album.service.ts`

m-m-mic commented on src/app/core/services/movies.service.ts:
Same as `album.service.ts`

m-m-mic commented on src/app/features/game/game.component.html:
Same as in `album.component`

m-m-mic commented on src/app/features/game/game.component.scss:
Same as in `album.component`

m-m-mic commented on src/app/features/game/game.component.ts:
Same as in `album.component`

m-m-mic commented on src/app/features/game/game.container.ts:
Same as in `album.component`

m-m-mic commented on src/app/features/game/game.form.ts:
Same as in `album.component`

m-m-mic commented on src/app/features/game/game.items.ts:
Same as in `album.component`

m-m-mic commented on src/app/features/games/games.component.html:
Same as in `albums.component`

m-m-mic commented on src/app/features/games/games.component.scss:
Same as in `albums.component`

m-m-mic commented on src/app/features/games/games.component.ts:
Same as in `albums.component`

m-m-mic commented on src/app/features/games/games.container.ts:
Same as in `albums.component`

m-m-mic commented on src/app/features/movie/movie.component.html:
Same as in `album.component`

m-m-mic commented on src/app/features/movie/movie.component.scss:
Same as in `album.component`

m-m-mic commented on src/app/features/movie/movie.component.ts:
Same as in `album.component`

m-m-mic commented on src/app/features/movie/movie.container.ts:
Same as in `album.component`

m-m-mic commented on src/app/features/movie/movie.form.ts:
Same as in `album.component`

m-m-mic commented on src/app/features/movie/movie.items.ts:
Same as in `album.component`

m-m-mic commented on src/app/features/movies/movies.component.html:
Same as in `albums.component`

m-m-mic commented on src/app/features/movies/movies.component.scss:
Same as in `albums.component`

m-m-mic commented on src/app/features/movies/movies.component.ts:
Same as in `albums.component`

m-m-mic commented on src/app/features/movies/movies.container.ts:
Same as in `albums.component`

m-m-mic committed:
chore: remove unused asset

m-m-mic commented on src/app/core/components/album-card/album-card.component.html:
```diff
@@ -0,0 +1,12 @@
+    <span [ngClass]="{ bookmark: true, visible: data.saved_by.includes(uid) }">
+      <ng-icon name="ionBookmark" size="24" />
+    </span>
+    <img class="card-image" src="assets/no_cover_placeholder.jpg" alt="Cover" />
```
Currently just a placeholder.

m-m-mic commented on src/app/core/components/header-button/header-button.component.html:
```diff
@@ -0,0 +1,73 @@
+  styleClass="account-button-menu"
+  [popup]="true" />
+
+<button class="burger-menu" (click)="sidebarVisible = true">
```
I added a separate burger menu which appears when the screen width is below 786px. Might be cleaner to split the component into two at some point.

georgiee replied:
The "header-button" name of the component and the content is not very obvious and makes it hard to understand. Would be better to have two navigation components, a login button and collect all of that in a header component where I can see the entire header structure.

But good that you start with two different contents. People often force themselves to create a css-only solution with a single markup but you often gain very few by doing that. It's essential that this duplicate content is never reachable by any user or screen reader though. Never should there be two navigations at the same time.

m-m-mic commented on src/app/core/models/album.interface.ts:
```diff
@@ -0,0 +1,26 @@
+  saved_by: string[];
+}
+
+export interface AlbumWithId extends Album {
```
Has to be a separate interface as the id has to be required whenever it is needed

m-m-mic commented on src/app/core/services/users.service.ts:
```diff
@@ -12,21 +18,312 @@ export class UsersService {
+  }
+
+  // Movie Collection methods
+  getUserMovie(movieId: string) {
```
All following methods are just adjusted copies of the Games Collection methods

m-m-mic commented on src/app/features/album/album.component.html:
```diff
@@ -0,0 +1,74 @@
+    <span [ngClass]="{ bookmark: true, visible: userAlbumData?.in_collection }">
+      <ng-icon name="ionBookmark" size="24" />
+    </span>
+    <img src="assets/no_cover_placeholder.jpg" alt="Cover" />
```
Also just a placeholder currently

m-m-mic commented on src/app/features/album/album.component.ts:
```diff
@@ -0,0 +1,69 @@
+    this.albumForm = this.formBuilder.group(fillAlbumForm(this.userAlbumData));
+  }
+
+  addToCollection() {
```
Does it makes sense to move these methods to the container? Doing so resulted in an error that the component didn't have access to services.

m-m-mic commented on src/app/features/dashboard/dashboard.component.ts:
```diff
@@ -2,23 +2,43 @@ import { Component } from '@angular/core';
+  uid = '';
+  loading = true;
+
+  constructor(private usersService: UsersService) {
```
The dashboard is currently very basic and will be overhauled in a future PR.

m-m-mic committed:
fix: add missing required

georgiee commented on src/app/app.component.scss:
```diff
@@ -3,3 +3,9 @@
   flex-flow: column;
 }
+
+.app-content {
+  flex: 1;
+  display: flex;
+  flex-flow: column;
+}
```
So centralizing the css for the overall layout on the router level did work ? 🙌

georgiee commented on src/app/app.component.spec.ts:
```diff
@@ -1,18 +0,0 @@
-import { RouterTestingModule } from '@angular/router/testing';
-import { AppComponent } from './app.component';
-
-describe('AppComponent', () => {
```
thanks for removing that.

georgiee commented on src/app/core/components/album-card/album-card.component.html:
```diff
@@ -0,0 +1,12 @@
+<a class="card" [routerLink]="['/albums/' + data.id]">
```
I think you can use the array as such: `[routerLink]="['/albums', data.id]` instead of concatenation

georgiee commented on src/app/core/components/album-card/album-card.component.scss:
```diff
@@ -0,0 +1,80 @@
+:host {
```
good css names and structure. very clear to read 👍

georgiee commented on src/app/core/components/album-card/album-card.component.ts:
```diff
@@ -0,0 +1,19 @@
+  imports: [CommonModule, NgIconComponent, RouterLink],
+  templateUrl: './album-card.component.html',
+  styleUrls: ['./album-card.component.scss'],
+  viewProviders: [provideIcons({ ionBookmark })],
```
Oh nice viewProviders. Never had a chance to use it. Is this described in the docs of the ui lib ?

m-m-mic replied:
Yes, its how primeNG handles icons

georgiee commented on src/app/core/components/game-card/game-card.component.html:
```diff
@@ -0,0 +1,12 @@
+<a class="card" [routerLink]="['/games/' + data.id]">
+  <span class="card-image-container">
+    <span [ngClass]="{ bookmark: true, visible: data.saved_by.includes(uid) }">
```
` bookmark: true` should be a static class, shouldn't it ?

georgiee replied:
`data.saved_by.includes(uid)` I would have expected an already resolved "visible" or "isBookmarked" here instead of digging in the data structure inside the template

georgiee commented on src/app/core/components/game-card/game-card.component.scss:
```diff
@@ -0,0 +1,80 @@
+:host {
```
ah i see the repetition happening. excellent playground to refactor later and thanks for keeping that part out of this PR.

m-m-mic committed:
chore: update readme

m-m-mic committed:
fix: add redirects file for netlify deployment

georgiee commented on src/app/core/components/header-button/header-button.component.scss:
```diff
@@ -0,0 +1,120 @@
+  cursor: pointer;
+  transition: 0.2s;
+
+  @include breakpoint-max(786px) {
```
you mix breakpoint min and max.
It's better to decide for one direction. For the conditional one (not being the default) you would override.
Here the default would be display none and you assign display block through a breakpoint min then. Makes it easiert to follow as you don't have arbitrary directions of changes.

georgiee commented on src/app/core/components/header-button/header-button.component.scss:
```diff
@@ -0,0 +1,120 @@
+  background: var(--primary-600);
+  padding: 4px;
+  border-radius: 6px;
+  ng-icon {
```
I can imagine why you do this, so it's okay to involve an element selector 😬

georgiee commented on src/app/core/components/header-button/header-button.component.scss:
```diff
@@ -0,0 +1,120 @@
+}
+
+.account-icon {
+  background: var(--primary-600);
```
is that a color assigment? would be better to use `background-color` then

georgiee commented on src/app/core/components/movie-card/movie-card.component.ts:
```diff
@@ -0,0 +1,19 @@
+  styleUrls: ['./movie-card.component.scss'],
+  viewProviders: [provideIcons({ ionBookmark })],
+})
+export class MovieCardComponent {
```
ignore mostly duplicate code (as described by you too)

georgiee commented on src/app/core/components/tab-selection/tab-selection.component.html:
```diff
@@ -0,0 +1,32 @@
+<div class="tab-selection">
```
Although it resemble a tab-like navigation the much better fit is to name it the main navigation.
tabs do the same: links that change a specific part of the screen estate. But it's a more local, content-block like concept. I would never apply it to the page navigation

georgiee commented on src/app/core/layout/header/header.component.html:
```diff
@@ -1,4 +1,5 @@
+<app-header-button
   [isLoggedIn]="isLoggedIn"
   [menuItems]="menuItems"
-  [displayName]="displayName"></app-mmsai-account-button>
```
thanks for changing the component names🙏

georgiee commented on src/app/core/layout/header/header.component.ts:
```diff
@@ -2,31 +2,37 @@ import { Component, DestroyRef, inject, OnInit } from '@angular/core';
+  ) {
+    this.authenticationService
       .getUser()
       .pipe(takeUntilDestroyed(this.destroyRef))
```
Note: you don't need the destroy ref if initialized within the constructor

georgiee commented on src/app/core/layout/header/header.component.ts:
```diff
@@ -2,31 +2,37 @@ import { Component, DestroyRef, inject, OnInit } from '@angular/core';
   styleUrls: ['./header.component.scss'],
 })
-export class HeaderComponent implements OnInit {
-  authenticationService = inject(AuthenticationService);
-  store = inject(Store);
-  destroyRef = inject(DestroyRef);
```
why did you move that into the constructor? Just curious. Both ways are good the constructor approach is the more classic/conservative approach.

m-m-mic replied:
usually if I need a constructor for something I will also inject services through it instead of doing it separately

georgiee commented on src/app/core/layout/loading/loading.component.ts:
```diff
@@ -1,13 +1,24 @@
-  templateUrl: './loading.component.html',
-  styleUrls: ['./loading.component.scss']
+  template: `
+    <div
```
thanks for inlining the template. Makes sense here

georgiee commented on src/app/core/layout/loading/loading.component.ts:
```diff
@@ -1,13 +1,24 @@
 })
 export class LoadingComponent {
-
+  @Input() height = 250;
```
oh I see that you're using that property a lot.
But it's two sizes: 160 & 120. Is it necessary to have two sizes? If yes I would make it more semantic like "big" and "small" as the input and then map a value. But nothing to change now as it works and it's really a smallish comment.

Also when dealing with loading states it might be good to have a few or one global state in the store and one loader that you mount atop of the application to show it whenever needed. Either lock he app (if needed) or display a the loading indicator at some place where it's not affecting the user's action (like the loading bar from medium for example)

georgiee commented on src/app/core/layout/loading/loading.component.ts:
```diff
@@ -1,13 +1,24 @@
+      }">
+      <span
+        class="pi pi-spin pi-spinner"
+        style="font-size: 1.6rem; height: 1.6rem;"></span>
```
inline styles are bad 👹

georgiee commented on src/app/core/models/album.interface.ts:
```diff
@@ -0,0 +1,26 @@
+  notes?: string;
+}
+
+export interface CombinedAlbum {
+  general: AlbumWithId;
+  user: UserAlbum;
+}
```
so nifty. I don't have a better idea but I don't like the combined thing 😁 I will look at the places where it's used maybe there is more simple solution, maybe those information just don't belong together and our brain tries too hard

georgiee commented on src/app/core/services/albums.service.ts:
```diff
@@ -0,0 +1,88 @@
+import {
+  AngularFirestore,
+  AngularFirestoreCollection,
+} from '@angular/fire/compat/firestore';
```
compat means compatibility. why do use use the compat layer of firestore. Is there some newer approach that you missed?

The examples on https://github.com/angular/angularfire look different. There is no `AngularFirestore` but a `Firestore`. What's that about ? They also use `collectionData` which I can't find in your repo

m-m-mic replied:
When I started implementing firestore I used a tutorial that used the compat functionalities as the firestore documentation itself is quite sparse in some areas. I might refactor it in the future

georgiee commented on src/app/core/services/albums.service.ts:
```diff
@@ -0,0 +1,88 @@
+
+    if (!album || !authUser?.uid) {
+      throwError(() => new Error('Could not find album or user'));
+      return;
```
+ Love to see explicit errors
+ The `return` is obsolete because the function halts when the error is thrown
+ You throw with rxjs's `throwError`. I think you should use `throw` in most cases, especially outside of any rxjs context

georgiee commented on src/app/core/services/albums.service.ts:
```diff
@@ -0,0 +1,88 @@
+
+    const data = album;
+    data.saved_by.push(authUser.uid);
+    await this.albumsRef.doc(albumId).update(data);
```
Hmm. Why do you resolve that? you don't do anything with the value and promises are running without  resolving them. So this should work ?

Also you await when you use `addUser` (I think I saw that somwhere) which would return nothing. So you would remove the await there too.

And seeing you using await/asycn so extensively is great. Let's the the theory straight now:

You absolutely have to learn about the js event loop if you didn't do that yet. My two favourites:

+ https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/
+ https://www.youtube.com/watch?v=8aGhZQkoFbQ

georgiee commented on src/app/core/services/games.service.ts:
```diff
@@ -0,0 +1,87 @@
+
+@Injectable({ providedIn: 'root' })
+export class GamesService {
+  private gamesPath = '/games';
```
you could make it a local constant. You don't need to reserve a private instance variable if you never change it.

```
const FIREBASE_PATH = '/games'
@Injectable({ providedIn: 'root' })
export class GamesService {
    this.gamesRef = db.collection(FIREBASE_PATH);
```

georgiee commented on src/app/core/services/games.service.ts:
```diff
@@ -0,0 +1,87 @@
+  }
+
+  async removeFromUserCollection(gameId: string) {
+    await this.removeUser(gameId);
+    await this.usersService.removeGameFromCollection(gameId);
```
you don't return any value and you don't do anything with the data. 
```suggestion
    this.removeUser(gameId);
    this.usersService.removeGameFromCollection(gameId);
```

That should work, shouldn't it ?

georgiee commented on src/app/features/album/album.container.ts:
```diff
@@ -0,0 +1,40 @@
+  standalone: true,
+  imports: [AsyncPipe, AlbumComponent],
+  template: `
+    <app-album
+      [id]="albumId"
+      [albumData]="(albumData$ | async) ?? undefined"
+      [userAlbumData]="(userAlbumData$ | async) ?? undefined" />
+  `,
```
nice container/presentational. Hoep it makes sense?

georgiee commented on src/app/features/dashboard/dashboard.component.html:
```diff
@@ -1,8 +1,54 @@
+</ng-template>
+
+<h2>Movies</h2>
+<app-loading [height]="120" *ngIf="loading; else moviesCollection" />
+<ng-template #moviesCollection>
```
ah I see what you're doing with the loading spinner.
Before I saw it and the different heights and assumed it's a loding spinner per page but it's per component so it's good.

A good spot to refactor after the PR though 👍

georgiee commented on src/app/features/movie/movie.container.ts:
```diff
@@ -0,0 +1,40 @@
+    private usersService: UsersService,
+    private route: ActivatedRoute,
+  ) {
+    this.route.paramMap.pipe(takeUntilDestroyed()).subscribe((params) => {
+      const paramId = params.get('movieId');
+      if (paramId) {
+        this.movieId = paramId;
+        this.movieData$ = this.moviesService.get(this.movieId);
+        this.userMovieData$ = this.usersService.getUserMovie(this.movieId);
+      }
+    });
+  }
```
It feels super wrong to use the subscribe to override observables. You ignore the streaming nature of rxjs. api driven I would prefer something like the following as a beginning

In pseudo code:

```
  this.movieData$ = this.route.paramMap.pipe(switchMap(paramid => this.moviesService.get(paramid))
  this.userMovieData$ =this.route.paramMap.pipe(switchMap(paramid => this.moviesService.get(paramdId))
```

you don't even need a take until destroyed because async will handle unsubscribing from the sources


also there is a common pattern: do not use the subscribe but use a tap for side effects.

```
something.subscribe(value => console.log(value))
// better
something.pipe(
  tap(value => console.log(value)
).subscribe()
)
```

That way you can easily defer the subscription or combine it with other streams

```
strean1$ =  something.pipe(
  tap(value => console.log(value)
)

strean1$.pipe(whatever).subscribe()

```

georgiee commented on src/app/shared/styles/breakpoints.scss:
```diff
@@ -0,0 +1,11 @@
+@mixin breakpoint-min($query) {
+  @media only screen and (min-width: $query) {
+    @content;
+  }
+}
+
+@mixin breakpoint-max($query) {
+  @media only screen and (max-width: $query) {
+    @content;
+  }
+}
```
Good to have a util for breakpoints 👍

georgiee reviewed (approved):
Hey thanks. I tried it out. It picked up my previous login so I was ready to start.
Works like a charm. Love it!

I went through ALL changes and commented on a few things. Even though I went through all changes I ignored some duplicates and at some point I accelerated my review speed a little bit.

I hope I still caught some important things.
You get my approval because it works and it's the base for future changes. So you can't make things wrong anyway but maybe you want to consider my feedback.


I love your PR summary, made it really simple to grasp what's going on ✨

m-m-mic committed:
chore: minor refactoring, media item pages no longer use subscribe on route
